File: shared-native/bindings/napi/crypto.cpp
#include <napi.h>
#include "gradido_blockchain_core/const.h"
#include "gradido_blockchain_core/crypto/sign.h"
#include "gradido_blockchain_core/result.h"

namespace gradido::crypto {
    constexpr uint32_t MAX_DERIVATION_INDEX = 0x80000000 - 1;

    Napi::Value GenerateFromSeed(const Napi::CallbackInfo& info)
    {
        Napi::Env env = info.Env();
        if (info.Length() != 1) {
            Napi::TypeError::New(env, "Expected one argument: seed (Uint8Array)").ThrowAsJavaScriptException();
            return env.Null();
        }
        if (!info[0].IsBuffer()) {
            Napi::TypeError::New(env, "Expected first argument to be a Uint8Array").ThrowAsJavaScriptException();
            return env.Null();
        }

        Napi::Buffer<uint8_t> seed = info[0].As<Napi::Buffer<uint8_t>>();

        // Result buffer: 32 bytes seed + 32 bytes public key + 32 bytes chain code = 96 bytes
        auto resultBuffer = Napi::Buffer<uint8_t>::New(env, sizeof(grdc_sign_key_pair));

        grd_result result = grdc_sign_key_pair_generate_from_seed(
            reinterpret_cast<grdc_sign_key_pair*>(resultBuffer.Data()),
            seed.Data(),
            seed.Length()
        );

        if (result != GRD_SUCCESS) {
            Napi::Error::New(env, std::string("Failed to generate ed25519 key pair, result: ") + grd_result_to_string(result))
                .ThrowAsJavaScriptException();
            return env.Null();
        }

        return resultBuffer;
    }

    Napi::Value Derive(const Napi::CallbackInfo& info)
    {
        Napi::Env env = info.Env();
        if (info.Length() != 2) {
            Napi::TypeError::New(env, "Expected two arguments: parentKeyPair (Uint8Array of 96 bytes) and index (number)")
                .ThrowAsJavaScriptException();
            return env.Null();
        }
        if (!info[0].IsBuffer()) {
            Napi::TypeError::New(env, "Expected first argument to be a Uint8Array").ThrowAsJavaScriptException();
            return env.Null();
        }
        if (!info[1].IsNumber()) {
            Napi::TypeError::New(env, "Expected second argument to be a number").ThrowAsJavaScriptException();
            return env.Null();
        }

        Napi::Buffer<uint8_t> parentKeyPair = info[0].As<Napi::Buffer<uint8_t>>();
        if (parentKeyPair.Length() != sizeof(grdc_sign_key_pair)) {
            Napi::TypeError::New(
                env,
                "Expected parentKeyPair to be "
                + std::to_string(sizeof(grdc_sign_key_pair))
                + " Bytes, got "
                + std::to_string(parentKeyPair.Length())
            ).ThrowAsJavaScriptException();
            return env.Null();
        }

        uint32_t index = info[1].As<Napi::Number>().Uint32Value();

        if (index > MAX_DERIVATION_INDEX) {
            Napi::TypeError::New(env, "Max index value is: " + std::to_string(MAX_DERIVATION_INDEX) + ", but got: " + std::to_string(index))
                .ThrowAsJavaScriptException();
            return env.Null();
        }

        auto resultBuffer = Napi::Buffer<uint8_t>::New(env, sizeof(grdc_sign_key_pair));

        grd_result result = grdc_sign_key_pair_derive(
            reinterpret_cast<grdc_sign_key_pair*>(resultBuffer.Data()),
            reinterpret_cast<const grdc_sign_key_pair*>(parentKeyPair.Data()),
            index
        );

        if (result != GRD_SUCCESS) {
            Napi::Error::New(env, std::string("Failed to derive child sign key, result: ") + grd_result_to_string(result))
                .ThrowAsJavaScriptException();
            return env.Null();
        }

        return resultBuffer;
    }

    Napi::Value DeriveUuid(const Napi::CallbackInfo& info)
    {
        Napi::Env env = info.Env();
        if (info.Length() != 2) {
            Napi::TypeError::New(env, "Expected two arguments: parentKeyPair (Uint8Array of 96 bytes) and uuid (Uint8Array of 16 bytes)")
                .ThrowAsJavaScriptException();
            return env.Null();
        }
        if (!info[0].IsBuffer()) {
            Napi::TypeError::New(env, "Expected first argument to be a Uint8Array").ThrowAsJavaScriptException();
            return env.Null();
        }
        if (!info[1].IsBuffer()) {
            Napi::TypeError::New(env, "Expected second argument to be a Uint8Array").ThrowAsJavaScriptException();
            return env.Null();
        }

        Napi::Buffer<uint8_t> parentKeyPair = info[0].As<Napi::Buffer<uint8_t>>();
        if (parentKeyPair.Length() != sizeof(grdc_sign_key_pair)) {
            Napi::TypeError::New(
                env,
                "Expected parentKeyPair to be 96 Bytes, got "
                + std::to_string(parentKeyPair.Length())
                + " bytes"
            ).ThrowAsJavaScriptException();
            return env.Null();
        }

        Napi::Buffer<uint8_t> uuid = info[1].As<Napi::Buffer<uint8_t>>();
        if (uuid.Length() != UUID_BINARY_SIZE) {
            Napi::TypeError::New(
                env,
                "Expected a valid uuid (16 Bytes), got "
                + std::to_string(uuid.Length())
                + " bytes"
            ).ThrowAsJavaScriptException();
            return env.Null();
        }

        auto resultBuffer = Napi::Buffer<uint8_t>::New(env, sizeof(grdc_sign_key_pair));

        grd_result result = grdc_sign_key_pair_derive_uuid(
            reinterpret_cast<grdc_sign_key_pair*>(resultBuffer.Data()),
            reinterpret_cast<const grdc_sign_key_pair*>(parentKeyPair.Data()),
            uuid.Data()
        );

        if (result != GRD_SUCCESS) {
            Napi::Error::New(env, std::string("UUID derivation failed: ") + grd_result_to_string(result))
                .ThrowAsJavaScriptException();
            return env.Null();
        }

        return resultBuffer;
    }

    Napi::Value DeriveAccountFromCommunity(const Napi::CallbackInfo& info)
    {
        Napi::Env env = info.Env();
        if (info.Length() < 2 || info.Length() > 3) {
            Napi::TypeError::New(env, "Expected two or three arguments: communitySeed (Uint8Array of 32 bytes), userUuid (Uint8Array of 16 bytes), and optional accountNumber (number, default 1)")
                .ThrowAsJavaScriptException();
            return env.Null();
        }
        if (!info[0].IsBuffer()) {
            Napi::TypeError::New(env, "Expected first argument to be a Uint8Array").ThrowAsJavaScriptException();
            return env.Null();
        }
        if (!info[1].IsBuffer()) {
            Napi::TypeError::New(env, "Expected second argument to be a Uint8Array").ThrowAsJavaScriptException();
            return env.Null();
        }

        Napi::Buffer<uint8_t> communitySeed = info[0].As<Napi::Buffer<uint8_t>>();
        if (communitySeed.Length() != SIGN_SEED_SIZE) {
            Napi::TypeError::New(
                env,
                "Expected a valid community seed (32 Bytes), got "
                + std::to_string(communitySeed.Length())
                + " bytes"
            ).ThrowAsJavaScriptException();
            return env.Null();
        }

        Napi::Buffer<uint8_t> userUuid = info[1].As<Napi::Buffer<uint8_t>>();
        if (userUuid.Length() != UUID_BINARY_SIZE) {
            Napi::TypeError::New(
                env,
                "Expected a valid uuid (16 Bytes), got "
                + std::to_string(userUuid.Length())
                + " bytes"
            ).ThrowAsJavaScriptException();
            return env.Null();
        }

        uint32_t accountNumber = 1;
        if (info.Length() > 2 && !info[2].IsUndefined()) {
            if (!info[2].IsNumber()) {
                Napi::TypeError::New(env, "Expected third argument to be a number").ThrowAsJavaScriptException();
                return env.Null();
            }
            accountNumber = info[2].As<Napi::Number>().Uint32Value();
        }

        auto resultBuffer = Napi::Buffer<uint8_t>::New(env, sizeof(grdc_sign_key_pair));

        grd_result result = grdc_sign_key_pair_derive_account_from_community(
            reinterpret_cast<grdc_sign_key_pair*>(resultBuffer.Data()),
            communitySeed.Data(),
            userUuid.Data(),
            accountNumber
        );

        if (result != GRD_SUCCESS) {
            Napi::Error::New(env, std::string("Account derivation failed: ") + grd_result_to_string(result))
                .ThrowAsJavaScriptException();
            return env.Null();
        }

        return resultBuffer;
    }

} // namespace gradido::crypto

File: shared-native/bindings/bun/crypto.ts
/**
 * C function declarations for grdc_sign are found in
 * {@link include/gradido_blockchain_core/crypto/sign.h}
 */

import { ptr } from 'bun:ffi'
import {
  blockchain_core,
  SIGN_CHAIN_CODE_SIZE,
  SIGN_PUBLIC_KEY_SIZE,
  SIGN_SEED_SIZE,
  UUID_BINARY_SIZE,
} from './library'

const MAX_DERIVATION_INDEX = 0x80000000 - 1
const KEY_PAIR_SIZE = SIGN_SEED_SIZE + SIGN_PUBLIC_KEY_SIZE + SIGN_CHAIN_CODE_SIZE

export function signKeyPairGenerateFromSeed(seed: Uint8Array): Uint8Array {
  const resultBuffer = new Uint8Array(KEY_PAIR_SIZE)
  const resultBufferPtr = ptr(resultBuffer)
  const seedPtr = ptr(seed)
  const result = blockchain_core.symbols.grdc_sign_key_pair_generate_from_seed(
    resultBufferPtr,
    seedPtr,
    seed.length,
  )
  if (result !== 0) {
    throw new Error(
      `[signKeyPairGenerateFromSeed] Failed to generate ed25519 key pair, result: ${blockchain_core.symbols.grd_result_to_string(result)}`,
    )
  }
  return resultBuffer
}

export function signKeyPairDerive(parentKeyPair: Uint8Array, index: number): Uint8Array {
  if (parentKeyPair.length !== KEY_PAIR_SIZE) {
    throw new Error(
      `[signKeyPairDerive] Expected parentKeyPair to be ${KEY_PAIR_SIZE} Bytes, got ${parentKeyPair.length}`,
    )
  }
  if (index > MAX_DERIVATION_INDEX) {
    throw new Error(
      `[signKeyPairDerive] Max index value is: ${MAX_DERIVATION_INDEX}, but got: ${index}`,
    )
  }
  const resultBuffer = new Uint8Array(KEY_PAIR_SIZE)

  const resultBufferPtr = ptr(resultBuffer)
  const parentKeyPairPtr = ptr(parentKeyPair)
  const result = blockchain_core.symbols.grdc_sign_key_pair_derive(
    resultBufferPtr,
    parentKeyPairPtr,
    index,
  )
  if (result !== 0) {
    throw new Error(
      `[signKeyPairDerive] Failed to derive child sign key, result: ${blockchain_core.symbols.grd_result_to_string(result)}`,
    )
  }
  return resultBuffer
}

export function signKeyPairDeriveUuid(parentKeyPair: Uint8Array, uuid: Uint8Array): Uint8Array {
  if (parentKeyPair.length !== KEY_PAIR_SIZE) {
    throw new Error(
      `[signKeyPairDeriveUuid] Expected parentKeyPair to be ${KEY_PAIR_SIZE} Bytes, got ${parentKeyPair.length} bytes`,
    )
  }
  if (uuid.length !== UUID_BINARY_SIZE) {
    throw new Error(
      `[signKeyPairDeriveUuid] Expected a valid uuid (${UUID_BINARY_SIZE} Bytes), got ${uuid.length} bytes`,
    )
  }
  const resultBuffer = new Uint8Array(KEY_PAIR_SIZE)

  const parentKeyPairPtr = ptr(parentKeyPair)
  const uuidPtr = ptr(uuid)
  const resultBufferPtr = ptr(resultBuffer)

  const result = blockchain_core.symbols.grdc_sign_key_pair_derive_uuid(
    resultBufferPtr,
    parentKeyPairPtr,
    uuidPtr,
  )
  if (result !== 0) {
    throw new Error(
      `[signKeyPairDeriveUuid] Failed to derive account key, result: ${blockchain_core.symbols.grd_result_to_string(result)}`,
    )
  }
  return resultBuffer
}

export function signKeyPairDeriveAccountFromCommunity(
  communitySeed: Uint8Array,
  userUuid: Uint8Array,
  accountNumber: number = 1,
): Uint8Array {
  if (communitySeed.length !== SIGN_SEED_SIZE) {
    throw new Error(
      `[signKeyPairDeriveAccountFromCommunity] Expected communitySeed to be ${SIGN_SEED_SIZE} Bytes, got ${communitySeed.length}`,
    )
  }
  if (userUuid.length !== UUID_BINARY_SIZE) {
    throw new Error(
      `[signKeyPairDeriveAccountFromCommunity] Expected a valid uuid (${UUID_BINARY_SIZE} Bytes), got ${userUuid.length} bytes`,
    )
  }
  if (accountNumber > MAX_DERIVATION_INDEX) {
    throw new Error(
      `[signKeyPairDeriveAccountFromCommunity] Max index value is: ${MAX_DERIVATION_INDEX}, but got: ${accountNumber}`,
    )
  }
  const resultBuffer = new Uint8Array(KEY_PAIR_SIZE)

  const communitySeedPtr = ptr(communitySeed)
  const userUuidPtr = ptr(userUuid)
  const resultBufferPtr = ptr(resultBuffer)

  const result = blockchain_core.symbols.grdc_sign_key_pair_derive_account_from_community(
    resultBufferPtr,
    communitySeedPtr,
    userUuidPtr,
    accountNumber,
  )
  if (result !== 0) {
    throw new Error(
      `[signKeyPairDeriveAccountFromCommunity] Failed to derive account key, result: ${blockchain_core.symbols.grd_result_to_string(result)}`,
    )
  }
  return resultBuffer
}
